src/DGtal/base/FunctorHolder.h

```diff
@@ -0,0 +1,291 @@
+/**
+ *  This program is free software: you can redistribute it and/or modify
+ *  it under the terms of the GNU Lesser General Public License as
+ *  published by the Free Software Foundation, either version 3 of the
+ *  License, or  (at your option) any later version.
+ *
+ *  This program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ **/
+
+#pragma once
+
+/**
+ * @file
+ * @author Roland Denis (\c denis@math.univ-lyon1.fr )
+ * CNRS, Institut Camille Jordan (UMR 5208), Université Lyon 1, France
+ *
+ * @date 2018/06/22
+ *
+ * This file is part of the DGtal library.
+ */
+
+#if defined(FunctorHolder_RECURSES)
+#error Recursive header files inclusion detected in FunctorHolder.h
+#else // defined(FunctorHolder_RECURSES)
+/** Prevents recursive inclusion of headers. */
+#define FunctorHolder_RECURSES
+
+#if !defined FunctorHolder_h
+/** Prevents repeated inclusion of headers. */
+#define FunctorHolder_h
+
+//////////////////////////////////////////////////////////////////////////////
+// Inclusions
+#include <iostream>
+#include <type_traits>
+#include <functional>
+#include <memory>
+//////////////////////////////////////////////////////////////////////////////
+
+namespace DGtal
+{
+namespace functors
+{
+
+//////////////////////////////////////////////////////////////////////////////
+// Implementation details
+namespace
+{
+
+  /// Invoke a callable object without dereferencing it before.
+  template <bool>
+  struct Invoker
+  {
+    template <typename F, typename... T>
+    static inline
+    auto apply(F && fn, T &&... args)
```

Comment:
Is that kind of "twice the same job for decltype and code" due to C++11 and solved in C++14 ?

Reply:
Yes: in C++14, return type can be automatically deduced from the return statement. If there is more than one return statement, they must be consistent.